app/core/src/main/java/com/fsck/k9/preferences/SettingsFileParser.kt

```diff
@@ -0,0 +1,441 @@
+                            skipElement()
+                        }
+                    }
+                    else -> {
+                        skipElement()
+                    }
+                }
+            }
+        }
+
+        return Imported(contentVersion, generalSettings, accounts)
+    }
+
+    private fun readFileFormatVersion(): Int {
+        val versionString = readAttribute(SettingsExporter.FILE_FORMAT_ATTRIBUTE)
+        return versionString.toIntOrNull()
+            ?: parserError("Invalid file format version: $versionString")
+    }
+
+    private fun readContentVersion(): Int {
+        val versionString = readAttribute(SettingsExporter.VERSION_ATTRIBUTE)
+        return versionString.toIntOrNull()?.takeIf { it > 0 }
+            ?: parserError("Invalid content version: $versionString")
+    }
+
+    private fun readGlobalSettings(): ImportedSettings? {
+        return readSettingsContainer()
+    }
+
+    private fun readSettingsContainer(): ImportedSettings? {
+        val settings = mutableMapOf<String, String>()
+
+        readElement { eventType ->
+            if (eventType == XmlPullParser.START_TAG) {
+                when (pullParser.name) {
+                    SettingsExporter.VALUE_ELEMENT -> {
+                        val key = readAttribute(SettingsExporter.KEY_ATTRIBUTE)
+                        val value = readText()
+
+                        if (settings.containsKey(key)) {
+                            Timber.w("Already read key \"%s\". Ignoring value \"%s\"", key, value)
+                        } else {
+                            settings[key] = value
+                        }
+                    }
+                    else -> {
+                        skipElement()
+                    }
+                }
+            }
+        }
+
+        return if (settings.isNotEmpty()) {
+            ImportedSettings(settings)
+        } else {
+            null
+        }
+    }
+
+    private fun readAccounts(): Map<String, ImportedAccount>? {
+        val accounts = mutableMapOf<String, ImportedAccount>()
+
+        readElement { eventType ->
+            if (eventType == XmlPullParser.START_TAG) {
+                when (pullParser.name) {
+                    SettingsExporter.ACCOUNT_ELEMENT -> {
+                        val account = readAccount()
+
+                        if (account == null) {
+                            // Do nothing - readAccount() already logged a message
+                        } else if (!accounts.containsKey(account.uuid)) {
+                            accounts[account.uuid] = account
+                        } else {
+                            Timber.w("Duplicate account entries with UUID %s. Ignoring!", account.uuid)
+                        }
+                    }
+                    else -> {
+                        skipElement()
+                    }
+                }
+            }
+        }
+
+        return accounts.takeIf { it.isNotEmpty() }
```

Comment:
Could be
```suggestion
        return if (accounts.isNotEmpty()) accounts else null
```

Reply:
I think once one got used to `takeIf`, that variant is easier to read and write.